dagstore_gc.go

```diff
@@ -73,6 +74,80 @@ func (d *DAGStore) gc(resCh chan *GCResult) {
 	}
 }
 
+// gc performs DAGStore GC. Refer to DAGStore#GC for more information.
+//
+// The event loops gives it exclusive execution rights, so while GC is running,
+// no other events are being processed.
+func (d *DAGStore) lruGC(target float64) {
+	// sort shards by LRU
+	d.lk.RLock()
+	var reclaim []*Shard
+	for _, s := range d.shards {
+		s.lk.RLock()
+		reclaim = append(reclaim, s)
+		s.lk.RUnlock()
+	}
+	d.lk.RUnlock()
+
+	sort.Slice(reclaim, func(i, j int) bool {
+		reclaim[i].lk.RLock()
+		defer reclaim[i].lk.RUnlock()
+
+		reclaim[j].lk.RLock()
+		defer reclaim[j].lk.RUnlock()
+
+		return reclaim[i].lastAccessedAt.Before(reclaim[j].lastAccessedAt)
+	})
+
+	// TODO What should we do about shards that are currently being initialised/acquired ?
+	// Removing the transient here will fail those operations
+	// The problem is those operations happen async outside the event loop.
+	// I think as a start, we should only enable automated LRU GC if client
+	// is okay with ongoing ops failing because of it i.e. "forced" mode.
+	// If client is NOT okay with the "forced" mode, client can fallback to the existing safe GC mechanism
+	// where we do NOT mess with shards that are being initialised or served.
+
+	// attempt to delete transients of reclaimed shards and free up required amount space.
+	for _, s := range reclaim {
+		// break when transient directory achieves required size
+		size, err := d.transientDirSize()
+		if err != nil {
+			continue
+		}
+		if float64(size) < target {
+			break
+		}
+
+		// only read lock: we're not modifying state, and the mount has its own lock.
+		// Is it okay to have a 'force' mode where it is okay to remove those transients and let the ops fail ?
+		s.lk.RLock()
+		err = s.mount.DeleteTransient()
+		if err != nil {
+			log.Warnw("failed to delete transient", "shard", s.key, "error", err)
+		}
+
+		// flush the shard state to the datastore.
+		if err := s.persist(d.ctx, d.config.Datastore); err != nil {
+			log.Warnw("failed to persist shard", "shard", s.key, "error", err)
+		}
```

Comment:
Why do this for every shard? Can't we do it once after the loop ends?

Reply:
(Need to ensure no corruption in case of an abrupt restart/interruption; however, that interruption can happen in the current code anyway, so we already need to be prepared for it)